drivers: flash: Add API for retrieve driver write-block-size requirement

nvlsianpu opened the pull request:
This path introduce API for retrieve minimum write-block-size supported by the flash driver.
This value can defer from hardware alignment requirement (as for nrf5x is).
It is not expected that this value changes in run-time so constant value was chosen.

As the driver has certain requirement for alignment
on the writing, it is necessary to export this value for upper modules which
needs to know the write-block-size (for instance NFFS needs this).

Appropriate values was specified for nRF5x, STM32,  w25qxxdv and QMSI SoCs.

nvlsianpu committed:
drivers: flash: Add API for retrieving a supported write-block-size

This path introduce API for retrieving a minimum write-block-size
supported by the flash driver.
This value can differ from the hardware alignment requirement
(as it does for nRF5x).

As the driver has a certain requirement for alignment
when writing, it is necessary to export this value for upper modules
which need to know the write-block-size (for instance, NFFS needs this).

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@carlescufi @nashif: This is proposition. As Kconfig doesn't have keyword for select integer value i use exact `range`. If you know better method for that - pleas let me know about.

nashif commented on drivers/flash/Kconfig:
```diff
@@ -162,4 +162,15 @@ config SOC_FLASH_MCUX_DEV_NAME
+	depends on FLASH
+	help
+	 The minimum write block size supported by the SoC flash driver.
+
```
where is this being used? The dependency should be done using depends on, and not on the prompt, or are you trying to do something else here?

nvlsianpu replied:
Nor `depends on FALSH && (SOC_FLASH_NRF5 || SOC_FLASH_STM32 || SOC_FLASH_QMS)` neither `depends on (SOC_FLASH_NRF5 || SOC_FLASH_STM32 || SOC_FLASH_QMS)` dose not work: config dispersing. For construction above this config is added in `flash hardware support menu` as expected.

carlescufi commented:
@andrzej-kaczmarek could you take a look?

andrzej-kaczmarek commented on drivers/flash/Kconfig:
```diff
@@ -162,4 +162,14 @@ config SOC_FLASH_MCUX_DEV_NAME
+	SOC_FLASH_QMSI)
+	range 1 1 if SOC_FLASH_NRF5
+	range 4 4 if SOC_FLASH_STM32
+	range 4 4 if SOC_FLASH_QMSI
```
using conditional defaults would be more readable

nvlsianpu replied:
Using keyword 'range' is only way to force the value. So I will keep this as is.

andrzej-kaczmarek commented on drivers/flash/Kconfig:
```diff
@@ -162,4 +162,14 @@ config SOC_FLASH_MCUX_DEV_NAME
 
+config SOC_FLASH_WRITE_BLOCK_SIZE
+	int "write block size (fixed)"
+	depends on FLASH && (SOC_FLASH_NRF5 || SOC_FLASH_STM32 ||\
```
I'd make it only depends on FLASH - just add default 0 or sth in case no "compatible" driver is enabled

galak commented:
Why aren't we generating this from DT?  Not sure I understand why we need a Kconfig for this.

nvlsianpu commented:
@galak This is software (SW) implementation related value. DT is defining hardware (HW) related things.

For instance on nRF5x SoCs DT is defining HW alignment of 4 B, and this patch introduce SW alignment of 1 B (thanks to read-modify implementation of the nrf5 driver).

carlescufi commented:
@galak to elaborate on what @nvlsianpu already said: we discussed this on IRC with @agross-linaro and the idea is that one thing is what the hardware supports and another is what the driver actually implements, which is something we need to expose to the filesystem sitting above it. So, in the case of Nordic ICs, the hardware requires 4-byte aligned 32-bit accesses to flash, but the driver is capable of doing any sort of unaligned access of any size thanks to the logic present in the software driver.

galak commented:
@nvlsianpu if this is what the driver supports and not a hw thing we need to rename the option, since calling it SOC_FLASH_... makes it seem like a HW setting.

nvlsianpu commented:
It is also possible to include proposed write-block-size requirement #define using header file included by flash.h header (or add this within flash.h) instead of using the kconfig.

I just want to signal another passibility. Personally I like Kconfig because it helps to keep user aware of settings.

galak commented on drivers/flash/Kconfig:
```diff
@@ -162,4 +162,15 @@ config SOC_FLASH_MCUX_DEV_NAME
 	help
 	  Specify the device name for the flash driver.
 
+config DRV_SOC_FLASH_WRITE_BLOCK_SIZE
```
config DRIVER_FLASH_WRITE_BLOCK_SIZE

galak commented on drivers/flash/Kconfig:
```diff
@@ -162,4 +162,15 @@ config SOC_FLASH_MCUX_DEV_NAME
 	  Specify the device name for the flash driver.
 
+config DRV_SOC_FLASH_WRITE_BLOCK_SIZE
+	int "write block size of the SoC driver (fixed)"
```
int "Minimum write block size supported by the driver"

(not sure what you mean by 'fixed')

galak commented on drivers/flash/Kconfig:
```diff
@@ -162,4 +162,15 @@ config SOC_FLASH_MCUX_DEV_NAME
+	range 4 4 if SOC_FLASH_QMSI
+	range -1 -1
+	help
+	 The minimum write block size supported by the SoC flash driver.
```
is this in bytes? (if so make that clear in the comments).  Also remove 'SoC' here.

nvlsianpu replied:
Are you surre? It is possible to have enabled SPI flash driver and SoC flash driver simultaneously.

galak replied:
Ok, so another reason to move this to the flash driver api, since we can support multiple different flash drivers.

nvlsianpu replied:
This will make impossible compile time optimization of a module which will use the driver. And there is always only one SoC flash driver.

galak replied:
how significant is such optimization really going to be?  I know there was some mention of some filesystem code needing this, can we see what that could that would use this look like.

Seems like we can have systems that have both internal and external flash that would need this option set, and possibly set different for each flash interface/driver.

carlescufi replied:
NFFS currently requires 1-byte granular unaligned access to flash. Starting there there is a problem today because we don't have a way of communicating that to NFFS today so it can assert or fail to initialize.

galak reviewed (changes requested):
Do we expect that someone will actually change this value?

I think it might be better to just add this to flash_driver_api is data that the driver sets.

carlescufi commented:
@galak having this as a compile-time option is important, so when in the future we add NFFS block size as an option it can pick it up from there. Having it in the driver itself would force us to include all code paths in the NFFS build and then decide at runtime. This is not ideal, but it yields the most optimized code and so I think it's worth it.

nvlsianpu committed:
drivers: flash: Add write-block-size values for SoCs

Write-block-size values were filled for nRF5, STM32, QMSI, MCUX and
w25qxxdv devices.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
sample: drivers: soc_flash_nrf5: Test of write-block-size get API

Added example code of using the new API for retrieving write-block-size.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@galak, @MaureenHelm Can you help with mcux write-block-size - I saw loot of conditionals in drivers implementations.

carlescufi reviewed (dismissed):
LGTM, someone should check the ST, QMSI and W25Q numbers

erwango commented on drivers/flash/flash_stm32f3x.c:
```diff
@@ -129,6 +129,7 @@ static const struct flash_driver_api flash_stm32_api = {
 	.write = flash_stm32_write,
 	.erase = flash_stm32_erase,
 	.write_protection = flash_stm32_protection_set,
+	.write_block_size = 1,
```
block size is half word on STM32F3

carlescufi replied:
@erwango so 2 bytes?

erwango replied:
@carlescufi: yes, 2. Otherwise it's good

nvlsianpu replied:
stm32f3 driver supports unaligned write, see: https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/flash/flash_stm32f3x.c#L93

nvlsianpu replied:
@erwango if it's good then please approve this change.

erwango replied:
@nvlsianpu: Well, looking into it, driver actually takes into account unaligned writes. Problem is that last leading byte of an unaligned write will be also written to 0x00 (data & 0x00FF operation on the remainder). And then this byte will no longer be writeable.
So write_block_size is indeed 1, but driver should be updated.

nvlsianpu replied:
@erwango So this driver has the bug.
I belief fixing of the driver is not in scope of this PR.

erwango replied:
@nvlsianpu: Sure. I'll take care of it

erwango replied:
@nvlsianpu: While correcting the issue, I had some testing and it appears that any attempt to write unaligned byte (of half-word) will end up in bus fault. So, even if driver allows to write single byte, STM32F3 flash does not allow unaligned writes.
To be transparent to the driver user, I think it would be better stating that write_block_size is 2

nvlsianpu replied:
Thx, I see - the driver allows to write any amount of data but it is limited to start writing for half-word aligned address.

erwango reviewed (changes requested):
From STM32 point of view, write block size

MaureenHelm commented:
> @galak, @MaureenHelm Can you help with mcux write-block-size - I saw loot of conditionals in drivers implementations.

Use the macro `FSL_FEATURE_FLASH_PFLASH_BLOCK_WRITE_UNIT_SIZE`

galak commented on include/flash.h:
```diff
@@ -207,6 +208,24 @@ int flash_get_page_info_by_idx(struct device *dev, u32_t page_index,
+ *  @brief  Get the minimum write block size supported by the driver
+ *
+ *  The Write block size supported by the driver might defer from the write
+ *  block size of memory used because the driver might implements write-modifi
```
write-modify?

galak reviewed (changes requested):
If we can update the MCUX driver I think this looks good and we can merge it (also slight typo in a comment)

nvlsianpu commented:
@galak updated
